src/providers/FusionAuthProvider.tsx

```diff
@@ -1,5 +1,108 @@
-import React from 'react';
+import React, { useCallback, useContext, useMemo } from 'react';
+import { TextEncoder } from 'util';
 
-export const FusionAuthProvider: React.FC = () => {
-    return <div>Hello</div>;
+export interface IFusionAuthContext {
+    login: (redirectURI: string, state: string) => void;
+}
+
+export const FusionAuthContext = React.createContext<IFusionAuthContext>({
+    login: () => {},
+});
+
+interface Props {
+    baseURL: string;
+    clientID: string;
+    scope: string;
+    children: React.ReactNode;
+}
+
+export const FusionAuthProvider: React.FC<Props> = ({
+    baseURL,
+    clientID,
+    scope,
+    children,
+}) => {
+    const login = useCallback(
+        async (redirectURI: string, state: string) => {
+            const fullURL = await generateURL(
+                FunctionType.login,
+                baseURL,
+                clientID,
+                scope,
+                redirectURI,
+                state,
+            );
+            window.location.assign(fullURL);
+        },
+        [baseURL, clientID, scope],
+    );
+
+    const providerValue = useMemo(
+        () => ({
+            login,
+        }),
+        [login],
+    );
+
+    return (
+        <FusionAuthContext.Provider value={providerValue}>
+            {children}
+        </FusionAuthContext.Provider>
+    );
 };
+
+export const useFusionAuthContext = () => useContext(FusionAuthContext);
+
+enum FunctionType {
+    login = 'authorize',
+    logout = 'logout',
+    register = 'register',
+}
+
+async function generateURL(
+    functionType: FunctionType,
+    baseURL: string,
+    clientID: string,
+    scope: string,
+    redirectURI: string,
+    state: string,
+) {
+    let fullURL = baseURL;
+    fullURL += `/${functionType}?`;
+    fullURL += `client_id=${clientID}&`;
+    fullURL += `scope=${scope}&`;
+    fullURL += `response_type=code&`;
+    fullURL += `redirect_uri=${redirectURI}&`;
+    fullURL += `code_challenge=${await generatePKCE()}&`;
+    fullURL += `code_challenge_method=S256&`;
+    fullURL += `state=${generateRandomString()}:${state}`;
+
+    return fullURL;
```

Comment:
Let's utilize [URLSearchParams](https://stackoverflow.com/a/52028292) to build the query params, something like:

```suggestion
    const queryString = new URLSearchParams({
        client_id: clientID,
        ...
    });
   
    return `${baseUrl}/${functionType}?${queryString}`;
```